docs/source/reference/oauth.md

```diff
@@ -0,0 +1,217 @@
+2. There are no credentials, so the client _redirects_ the browser to an "authorize" page on the oauth _provider_ with some extra information:
+   - the oauth **client id** of the client itself
+   - the **redirect uri** to be redirected back to after completion
+   - the **scopes** requested, which the user should be presented with to confirm.
+     This is the "X would like to be able to Y on your behalf. Allow this?" page you see on all the "Login with ..." pages around the Internet.
+3. During this authorize step,
+   the browser must be _authenticated_ with the provider.
+   This is often already stored in a cookie,
+   but if not the provider webapp must begin its _own_ authentication process before serving the authorization page.
+4. After the user tells the provider that they want to proceed with the authorization,
+   the provider records this authorization in a short-lived record called an **oauth code**.
+5. Finally,
+   the oauth provider redirects the browser _back_ to the oauth client's "redirect uri"
+   (or "oauth callback uri"),
+   with the oauth code in a url parameter.
+
+#### State after redirect
+At this point:
+
+- The browser is authenticated with the _provider_
+- The user's authorized permissions are recorded in an _oauth code_
+- The _provider_ knows that the given oauth client's requested permissions have been granted, but the client doesn't know this yet.
+- All requests so far have been made directly by the browser.
+  No requests have originated at the client or provider.
+
+#### OAuth Client Handles Callback Request
+Now we get to finish the OAuth process.
+Let's dig into what the oauth client does when it handles
+the oauth callback request with the
+
+- The OAuth client receives the _code_ and makes an API request to the _provider_ to exchange the code for a real _token_.
+  This is the first direct request between the OAuth _client_ and the _provider_.
+- Once the token is retrieved, the client _usually_
+  makes a second API request to the _provider_
+  to retrieve information about the owner of the token (the user)
+- Finally, the oauth client stores its own record that the user is authorized in a cookie.
+  This could be the token itself, or any other appropriate representation of successful authentication.
+
+_phew_
+
+So that's _one_ OAuth process.
+
+## Full sequence of OAuth in JupyterHub
+
+Let's go through the above oauth process in Jupyter,
+with specific examples of each HTTP request and what information is contained.
+
+Our starting point:
+
+- a user's single-user server is running. Let's call them `danez`
+- jupyterhub is running with GitHub as an oauth provider,
+- Danez has a fresh browser session with no cookies yet
+
+First request:
+
+- browser->single-user server running JupyterLab or Jupyter Classic
+- `GET /user/danez/notebooks/mynotebook.ipynb`
+- no credentials, so client starts oauth process with JupyterHub
+- response: 302 redirect -> `/hub/api/oauth2/authorize`
```

Comment:
`no credentials, so client starts oauth process with JupyterHub` : In this sentence (and in the other of this explanation), the word `client` may be ambiguous. Does it mean the effective end-user, or is it the oauth client ?

Reply:
`client` is defined above: - **client** An entity that requests OAuth tokens on a user's behalf.
  JupyterHub _services_ or single-user _servers_ are OAuth clients of the JupyterHub _provider_.
  When OAuthenticator is used, JupyterHub is itself also an OAuth _client_ for the external oauth _provider_, e.g. GitHub.

Sticking to that definition, some of my above comments are answered and the current text is OK. It is just a bit hard to convert a physical user to those OAuth clients. Maybe an introductory text that stresses on that aspect would help.

Reply:
Since it's an oauth keyword, maybe we should keep it bold and/or capitalized ot make it clear that client is a specific term meaning "the application using jupyterhub for authentication, registered with a client id" - i.e. the single-user server in this case.

I'll give the terminology intro another pass with this in mind.